App/Program.cs

```diff
@@ -13,26 +14,58 @@ class Program
     {
         static void Main(string[] args)
         {
-#if NETCOREAPP
-            var alc = new MyBindingRedirectAssemblyLoadContext();
-            var a = alc.LoadFromAssemblyPath(typeof(Program).Assembly.Location);
-            var t = a.GetType(typeof(Program).FullName);
-            var t_self = typeof(Program);
-            t.GetMethod(nameof(DoMsgPackStuff), BindingFlags.NonPublic | BindingFlags.Static).Invoke(null, null);
-#else
             DoMsgPackStuff();
-#endif
         }
 
         private static void DoMsgPackStuff()
         {
-            const string message = "hi";
-            Console.WriteLine($"V2 serialized: {Convert.ToBase64String(MsgPackV2Consumer.Serialize(message))}");
-            Console.WriteLine($"V1 serialized: {Convert.ToBase64String(MsgPackV1Consumer.Serialize(message))}");
+            Poco message = new() { Message = "hi" };
+
+            var v2data = MsgPackV2Consumer.Serialize(message);
+            Console.WriteLine($"V2 serialized: {Convert.ToBase64String(v2data)}");
+            var v2message = MsgPackV2Consumer.Deserialize(typeof(Poco), v2data);
+
+            var v1data = MsgPackV1Consumer.Serialize(message);
+            Console.WriteLine($"V1 serialized: {Convert.ToBase64String(v1data)}");
+            var v1message = MsgPackV1Consumer.Deserialize(typeof(Poco), v1data);
         }
     }
 
-#if NETCOREAPP
+    public class Poco
+    {
+        public string Message { get; set; }
+    }
+
+#if NET7_0
+    internal class MsgPackV1Consumer
+    {
+        private static MsgPackV1Consumer _instance;
+        public static MsgPackV1Consumer Instance
+        {
+            get
+            {
+                _instance ??= new MsgPackV1Consumer();
+                return _instance;
+            }
+        }
+
+        private readonly Type _sideLoadedMsgPackV1Consumer;
+        private readonly MethodInfo _serializeMethod;
+        private readonly MethodInfo _deserializeMethod;
+
+        public MsgPackV1Consumer()
+        {
+            var assemblyLoadContext = new MyBindingRedirectAssemblyLoadContext();
+            var assembly = assemblyLoadContext.LoadFromAssemblyPath(typeof(MsgPackV1.MsgPackV1Consumer).Assembly.Location);
+            _sideLoadedMsgPackV1Consumer = assembly.GetType(typeof(MsgPackV1.MsgPackV1Consumer).FullName);
+            _serializeMethod = _sideLoadedMsgPackV1Consumer.GetMethod(nameof(Serialize), new Type[] { typeof(string) });
+            _deserializeMethod = _sideLoadedMsgPackV1Consumer.GetMethod(nameof(Deserialize), new Type[] { typeof(Type), typeof(byte[]) });
+        }
```

Comment:
I'm pretty sure I can come up with a far more natural way of doing this nowadays that I've dealt with ALCs more. I'll try it out and possibly push a change to your PR.

Reply:
Please do. I didn't like where I landed with it but I considered it plumbing code that should only need to be written once. 

I'd love to see alternatives 🙂

Reply:
Alternative pushed.